enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,693 @@
+No other components talk to the CCM directly. The CCM coordinates with other components by updating resources using the KAS similar to other controllers.
+
+Q: CCM relation to other openshift components, such as SDN and storage? How a non-operational CCM will affect cluster health, which components will take a hit?
+
+- A: A couple of components rely on CCM at different time of the cluster lifetime:
+  - `kubelet` dependency requires removal of taints from newly created Nodes. Without CCM, new nodes will not become schedulable, existing nodes are unaffected.
+  - CCMs manage `LoadBalancer` `Service` resources in all in-tree cloud implementations, except vSphere, so their creation will not be possible while the CCM is down.
+  This will affect the [ingress-operator](https://github.com/openshift/cluster-ingress-operator).
+  Cloud networking across `Nodes` is [disabled](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/8be94db9da523152af2268bd7f891fe089a424eb/bindata/v4.1.0/config/defaultconfig.yaml#L8-L9) by default for all clouds, so the routes will not be provisioned for the Node by the `CCM` (which is done by `Routes()` method in the interface), and so not required for SDN functionality.
+  - Storage CSI migration and CSI provider is a requirement for each cloud provider.
+
+Q: Should we reuse the existing cloud provider config or generate a new one?
+
+- A: CCM config is backward compatible with the in-tree cloud provider, therefore we can reuse it.
+
+Q: How to migrate PVs created by the in-tree cloud provider?
+
+- A: [CSIMigration](https://kubernetes.io/blog/2019/12/09/kubernetes-1-17-feature-csi-migration-beta/) is the preferred option. It graduates in GA in 1.19 (1.20).
+
+Q: Does CCM provide an API? How is it hosted? HTTPS endpoint? Through load balancer?
+
+- A: Not at the moment. The CCM vendors part of the code serving cloud features from the [cloud-provider](https://github.com/kubernetes/cloud-provider). CCM provides a secure TCP connection on port [10258](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/ports.go#L22) and it is only used for  metrics and readiness/health endpoint.
+
+Q: What happens if the kcm leader has no access to ccm? Can it continue its work? Will it give up leadership?
+
+- A: KCM and CCM perform independently after the initial migration process. KCM will be unaffected by CCM being down after the initial migration. The CCM cluster operator will be responsible for transitioning to Unavailable or Degraded in a similar situation.
+
+Q: What are the thoughts about certificate management?
+
+- A: At present, only the GCP provider has any CSR approval mechanism built in. This functionality is used to approve `NodeClient` and `NodeServer` certificates for Nodes on GCP.
+This functionality should not be required as OpenShift handles certificate approval via the `cluster-machine-approver` component.
+
+Q: Does every node need a CCM?
```

Comment:
I don't see any mention of `cloud-node-controller` in the enhancement. Isn't that also needed for this transition?

Reply:
The `cloud-node-manager` (I think that's the same) is typically included in the CCM. AFAIK only Azure runs it as a separate DaemonSet (certainly for the platforms OCP supports today). I believe that is called out somewhere within this enhancement already.

Looking now it seems the note we had about running the DaemonSet for azure has gone, will make sure we add something back

Reply:
Added a minor comment describing `CNM` role in this question.